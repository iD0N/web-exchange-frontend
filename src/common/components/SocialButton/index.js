import React from 'react';
import PropTypes from 'prop-types';
import Icon from 'antd/lib/icon';

import Button from '../Button';

const SocialButton = ({ type, children, ...props }) => (
  <Button {...props} type={type}>
    <Icon type={type} theme={type === 'facebook' ? 'filled' : undefined} />
    {children}
  </Button>
);

SocialButton.propTypes = {
  type: PropTypes.oneOf(['google', 'facebook']),
  children: PropTypes.node.isRequired,
};

export default SocialButton;
