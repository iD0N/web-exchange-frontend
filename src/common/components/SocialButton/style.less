.ant-btn {
  &-facebook {
    .button-variant-primary(@white; @blue-facebook);
  }

  &-google {
    .button-variant-primary(@white; @orange-google);
  }
}

.social-buttons {
  .ant-btn {
    margin-bottom: @padding-md;
    display: flex;
    font-size: 15px;
    letter-spacing: 0;

    .anticon {
      height: @btn-height-base - 2px;
      font-size: 24px;
      svg {
        height: inherit;
      }
    }
  }
}
