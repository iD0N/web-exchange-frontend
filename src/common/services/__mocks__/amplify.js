const { getLoginUrl, configureAmplify } = require.requireActual('../amplify');
export { getLoginUrl, configureAmplify };
// TODO:
// when requiring actual & exporting mock, the function is not supplied in actual implementation which uses the module
export const handleAmplifyError = jest.fn();
