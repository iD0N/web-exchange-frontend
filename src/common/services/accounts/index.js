// TODO: move to other dir as this is an app service, not a common service
import { createSelector } from 'reselect';
import { all, call, put, select, takeLatest } from 'redux-saga/effects';

import { withApiCall } from '../spinner';

import {
  createActionType,
  createApiActionCreators,
  createReducer,
  REQUEST,
  SUCCESS,
} from '../../utils/reduxHelpers';
import { selectHasAppAccess, setTraderIdAction } from '../user';

import api, { apiCallIds } from './api';

/**
 * ACTION TYPES
 */
export const FETCH_ACCOUNTS = 'accounts/FETCH_ACCOUNTS';
export const CREATE_ACCOUNT = 'accounts/CREATE_ACCOUNT';
export const DELETE_ACCOUNT = 'accounts/DELETE_ACCOUNT';
export const INTERNAL_TRANSFER = 'accounts/INTERNAL_TRANSFER';
export const REBALANCE_USD = 'accounts/REBALANCE_USD';

/**
 * ACTIONS
 */
export const fetchAccountsAction = createApiActionCreators(FETCH_ACCOUNTS);
export const createAccountAction = createApiActionCreators(CREATE_ACCOUNT);
export const deleteAccountAction = createApiActionCreators(DELETE_ACCOUNT);
export const internalTransferAction = createApiActionCreators(INTERNAL_TRANSFER);
export const rebalanceUsdAction = createApiActionCreators(REBALANCE_USD);

/**
 * REDUCERS
 */
const initialState = {
  accounts: [],
};

const accounts = createReducer(initialState.accounts, {
  [FETCH_ACCOUNTS]: {
    [SUCCESS]: (state, accounts) => accounts,
  },
});

export default accounts;

/**
 * SELECTORS
 */

export const selectAccountsData = state => state.accounts;

export const selectUser = state => state.user;
export const selectTraderId = state => selectUser(state).traderId || (state.accounts[0] || {}).traderId;

export const selectAccounts = createSelector(
  selectAccountsData,
  selectTraderId,
  (accounts, currentId) =>
    accounts.map((account, index) => ({
      ...account,
      isSelectedTraderId: currentId ? account.traderId === currentId : index === 0,
      isPrimary: index === 0,
      displayName: `${index === 0 ? 'Main Account' : (account.alias || account.anonymousName)}`,
    }))
);

export const selectFirstAccount = createSelector(selectAccounts, accounts => accounts[0] || {});

export const selectAccountTraderId = createSelector(
  selectAccounts,
  selectTraderId,
  selectFirstAccount,
  (accounts, traderId, { traderId: defaultTraderId }) =>
    !!traderId && accounts.length > 0 ? traderId : defaultTraderId
);

export const selectAccount = createSelector(
  selectAccounts,
  selectAccountTraderId,
  (accounts, traderId) =>
    accounts.find(({ traderId: accountTraderId }) => accountTraderId === traderId) || {}
);

export const selectAccountAlias = createSelector(
  selectAccount,
  ({ alias, anonymousName, isPrimary }) => alias || (isPrimary ? undefined : anonymousName)
);

export const selectTokenBalances = state => selectAccount(state).tokenBalances.byId;
export const selectAffiliateIncome = createSelector(
  selectAccount,
  ({ usdBalanceByType }) => (!!usdBalanceByType && usdBalanceByType.affiliateIncome) || 0
);

export const selectAccountsLoaded = createSelector(selectAccounts, accounts => accounts.length > 0);

/**
 * SAGA HELPERS
 */
export function* updateAccountAlias(alias) {
  const traderId = yield select(selectAccountTraderId);

  yield call(api.updateAccountAlias, traderId, alias);
  yield call(fetchAccountsWithBalanceData);
}

/**
 * SAGAS
 */
function* fetchAccountsWithBalanceData() {
  const hasAppAccess = yield select(selectHasAppAccess);

  if (!hasAppAccess) {
    return;
  }

  const accountsResp = yield call(api.getAccounts);

  if (accountsResp && accountsResp.ok) {
    const { accounts } = accountsResp.data;

    try {
      const subaccountsData = yield all(
        accounts.map(({ traderId }) => call(api.getAccountData, traderId))
      );
      const idToData = subaccountsData.reduce(
        (map, data, index) => (data.ok ? { ...map, [accounts[index].traderId]: data.data } : map),
        {}
      );

      const result = accounts.map(account => ({
        ...account,
        ...(idToData[account.traderId] || {}),
      }));

      yield put(fetchAccountsAction.success(result));
    } catch (err) {
      yield put(fetchAccountsAction.success(accounts));
    }
  }
}

function* createAccount({ payload }) {
  const resp = yield call(api.createAccount, payload);

  if (resp.ok) {
    yield put(fetchAccountsAction.request());
  }
}

function* deleteAccount({ payload: traderId }) {
  const resp = yield call(api.deleteAccount, traderId);
  const currentId = yield select(selectTraderId);

  if (resp.ok) {
    if (currentId === traderId) {
      // deleted the active subaccount, toggle back to default
      yield put(setTraderIdAction(null));
      return;
    }
    yield put(fetchAccountsAction.request());
  }
}

function* internalTransfer({ payload }) {
  const resp = yield call(api.internalTransfer, payload);

  if (resp.ok) {
    yield put(fetchAccountsAction.request());
  }
}

function* rebalanceUsd() {
  const currentId = yield select(selectTraderId);
  yield call(api.rebalanceUsd, currentId);
}

export function* accountsSaga() {
  yield takeLatest(
    createActionType(FETCH_ACCOUNTS, REQUEST),
    withApiCall(apiCallIds.FETCH_ACCOUNTS, fetchAccountsWithBalanceData)
  );
  yield takeLatest(createActionType(CREATE_ACCOUNT, REQUEST), createAccount);
  yield takeLatest(createActionType(DELETE_ACCOUNT, REQUEST), deleteAccount);
  yield takeLatest(createActionType(INTERNAL_TRANSFER, REQUEST), internalTransfer);
  yield takeLatest(createActionType(REBALANCE_USD, REQUEST), rebalanceUsd);
}
