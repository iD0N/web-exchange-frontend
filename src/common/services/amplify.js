import Amplify, { Auth } from 'aws-amplify';
import qs from 'qs';
import { pascalCase } from 'change-case';

import { store } from '../../app/store/configureStore';
import config from '../../config';
import { AMPLIFY_ERROR_CODES } from '../enums';
import { silentReloginAction } from './user';
import {
  CognitoUser,
  CognitoUserSession,
  CognitoIdToken,
  CognitoRefreshToken,
  CognitoAccessToken,
} from 'amazon-cognito-identity-js';

import { logoutAction } from './user';
import i18n from './i18n';
import AlertService from './alert';

window.LOG_LEVEL = config().logLevel; // enable Amplify logger plus others

let amplify;

window.mobileUser = null;

export function configureAmplify() {
  // Custom method
  // nested in configure function, to not attach it in test env
  Auth.refreshSession = (cognitoUser, refreshToken) =>
    new Promise((resolve, reject) =>
      cognitoUser.refreshSession(refreshToken, (err, session) => {
        if (err) reject(err);
        resolve(session);
      })
    );

  const amplifyConfig = config().amplify;
  if (window.ReactNativeWebView) {
    // switch to mobile client
    amplifyConfig.userPoolWebClientId = amplifyConfig.userPoolMobileClientId;
  }

  amplify = Amplify.configure(amplifyConfig);

  //
  // The native mobile app calls this through a WebView injection
  //
  window.mobileLogin = async function(mobileSession) {
    const mobileRefreshToken = new CognitoRefreshToken({
      RefreshToken: mobileSession.refreshToken.token,
    });

    const localSession = new CognitoUserSession({
      IdToken: new CognitoIdToken({ IdToken: mobileSession.idToken.jwtToken }),
      RefreshToken: mobileRefreshToken,
      AccessToken: new CognitoAccessToken({ AccessToken: mobileSession.accessToken.jwtToken }),
    });

    const localUser = new CognitoUser({
      Username: mobileSession.accessToken.payload.username,
      Pool: Auth.userPool,
      Storage: Auth.userPool.storage,
    });
    localUser.setSignInUserSession(localSession);

    // pass along the device key (cached in storage and passed when refreshing session)
    localUser.deviceKey = mobileSession.accessToken.payload.device_key;
    localUser.cacheDeviceKeyAndPassword();

    window.mobileUser = localUser;

    try {
      // login failed?
      if (!(await Auth.currentAuthenticatedUser())) {
        console.error('mobile login failure');
        return;
      }
      store.dispatch(silentReloginAction());
    } catch (ex) {
      console.error(`mobile login excepction: ${ex}`);
    }
  };
}

export function getLoginUrl(provider) {
  const { userPoolWebClientId: clientId, oauth } = amplify;
  const { domain, redirectSignIn, responseType, scope } = oauth;

  const params = {
    redirect_uri: encodeURIComponent(redirectSignIn),
    response_type: responseType,
    client_id: clientId,
    scope: scope.join('+'),
    identity_provider: provider,
  };

  return `https://${domain}/oauth2/authorize?${qs.stringify(params, {
    encode: false,
  })}`;
}

export function handleAmplifyError(err) {
  const error = normalizeAmplifyError(err);

  if (
    [AMPLIFY_ERROR_CODES.NOT_AUTHENTICATED, AMPLIFY_ERROR_CODES.NO_CURRENT_USER].includes(
      error.code
    )
  ) {
    store.dispatch(logoutAction());
    return;
  }

  const key = `authErrors.${error.code}`;
  const message = i18n.exists(key) ? i18n.t(key) : error.message;
  AlertService.error(message);
}

function normalizeAmplifyError(error) {
  return typeof error === 'string'
    ? {
        code: pascalCase(error),
        message: error,
      }
    : error;
}
