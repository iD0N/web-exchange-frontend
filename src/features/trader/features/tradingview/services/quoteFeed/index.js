import moment from 'moment';
import BigNumber from 'bignumber.js';

import { isProd } from '../../../../../../config';
import apiClient from '../../../../../../common/services/apiClient';
import { CONTRACT_TYPE } from '../../../../../../common/enums';

import {
  transform,
  createQueryIntervals,
  generateGappedQuotes,
  MAX_BUCKETS,
  reBucket,
} from './utils';

const api = {
  // TODO: Public route, make request without token?
  getCandles: (contractCode, startTime, endTime, granularity, contractType) =>
    apiClient.get(`/contracts/${contractCode}/candles`, {
      params: {
        ...(isProd() && contractType !== CONTRACT_TYPE.SPOT
          ? { fieldName: 'mark_price' }
          : contractType === CONTRACT_TYPE.SPOT
          ? { fieldName: 'fair_price' }
          : {}),
        startTime,
        endTime,
        granularity: `${granularity}m`,
      },
      apiCallId: null,
    }),
};

export const multiQueryCandles = (contractCode, queries, granularity, contractType) =>
  new Promise(resolve => {
    if (!queries.length) {
      return resolve([]);
    }

    Promise.all(
      queries.map(({ startStr, endStr }) =>
        api.getCandles(contractCode, startStr, endStr, granularity, contractType).then(transform)
      )
    )
      .then(results => {
        //why need to use reduce() to pick out the quotes data?
        resolve(
          results.reduce((allQuotes, { ok, data: quotes } = {}) => {
            return ok && quotes.length ? [...allQuotes, ...quotes] : [...allQuotes];
          }, [])
        );
      })
      .catch(err => {
        console.log(err);
      });
  });

const quoteFeedService = ({ contractsMetadata }) => ({
  fetchInitialData: (contractCode, start, end, parameters) => {
    const startDate = start * 1000;
    const endDate = end * 1000;
    return new Promise((resolve, reject) => {
      if (!contractsMetadata[contractCode]) {
        return;
      }

      let needsRebucket = false;

      let { resolution: granularity } = parameters;
      if (
        granularity === 'D' ||
        granularity === 'M' ||
        granularity === 'W' ||
        Number(granularity) > 30
      ) {
        granularity = '30';
        needsRebucket = true;
      }

      const mListTime = moment(contractsMetadata[contractCode].listTime);
      const mStartDate = moment.max(moment(startDate), mListTime);

      const startIsAfterListTime = mStartDate.isAfter(mListTime);
      const queries = createQueryIntervals(mStartDate, endDate, granularity);

      multiQueryCandles(
        contractCode,
        queries,
        granularity,
        contractsMetadata[contractCode].type
      ).then(data => {

        let quotes = data.sort((a, b) => a.time - b.time);
        // .filter(({ close, indexPrice }) => !!close);

        if (needsRebucket) {
          quotes = reBucket(quotes, 1000 * 60 * 60 * 24);
        }
        resolve({
          quotes,
          moreAvailable: startIsAfterListTime,
        });
      });
    });
  },

  fetchPaginationData(contractCode, startDate, endDate, parameters, cb) {
    if (!contractsMetadata[contractCode]) {
      return;
    }
    const { period: granularity } = parameters;

    const { listTime } = contractsMetadata[contractCode];
    const startIsAfterListTime = moment(startDate).isAfter(moment(listTime));
    const queries = createQueryIntervals(startDate, endDate, granularity);

    multiQueryCandles(
      contractCode,
      [...queries],
      granularity,
      contractsMetadata[contractCode].type
    ).then(quotes =>
      cb({
        quotes: quotes.length
          ? quotes
          : generateGappedQuotes(startIsAfterListTime ? startDate : listTime),
        moreAvailable:
          startIsAfterListTime ||
          quotes.length ===
            BigNumber(MAX_BUCKETS)
              .multipliedBy(queries.length)
              .toNumber(),
      })
    );
  },
});

export default quoteFeedService;
