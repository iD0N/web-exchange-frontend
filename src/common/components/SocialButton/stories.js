import React from 'react';
import { storiesOf } from '@storybook/react';

import SocialButton from './';

storiesOf('SocialButton', module)
  .add('Google', () => <SocialButton type="google">Sign in</SocialButton>)
  .add('Facebook', () => <SocialButton type="facebook">Sign in</SocialButton>);
