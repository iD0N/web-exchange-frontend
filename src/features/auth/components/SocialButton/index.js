import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Trans } from 'react-i18next';
import { Cookies, withCookies } from 'react-cookie';

import {} from '../../../../config';
import { getLoginUrl } from '../../../../common/services/amplify';
import SocialButton from '../../../../common/components/SocialButton';

class EnhancedSocialButton extends Component {
  static propTypes = {
    provider: PropTypes.oneOf(['Google', 'Facebook']),
    isSignUp: PropTypes.bool,
    cookies: PropTypes.instanceOf(Cookies).isRequired,
  };

  static defaultProps = {
    isSignUp: false,
  };

  handleRedirect = e => {
    e.preventDefault();
    try {
      Object.keys(this.props.allCookies)
        .filter(name => name.match('CognitoIdentity'))
        .forEach(name => {
          this.props.cookies.remove(name, { path: '/', domain: '.crypto.io' });
        });
    } catch (err) {}
    setTimeout(() => {
      window.location.href = getLoginUrl(this.props.provider);
    }, 50);
  };

  render() {
    const { provider, cookies, allCookies, isSignUp, ...props } = this.props;
    return (
      <SocialButton
        {...props}
        type={provider.toLowerCase()}
        href={getLoginUrl(provider)}
        onClick={this.handleRedirect}
      >
        <span>
          {isSignUp ? (
            <Trans i18nKey="socialButton.signUp.label">Sign Up with {provider}</Trans>
          ) : (
            <Trans i18nKey="socialButton.logIn.label">Login with {provider}</Trans>
          )}
        </span>
      </SocialButton>
    );
  }
}

export default withCookies(EnhancedSocialButton);
