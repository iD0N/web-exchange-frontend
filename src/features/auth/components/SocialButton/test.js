/* eslint-disable import/first */
jest.mock('react-i18next');
jest.mock('../../../../common/services/i18n');
jest.mock('../../../../common/services/amplify');

import React from 'react';
import renderer from 'react-test-renderer';

import { configureAmplify } from '../../../../common/services/amplify';

import SocialButton from '../SocialButton';

describe('Social buttons', () => {
  beforeAll(() => {
    configureAmplify();
  });

  it('provides Google sign in', () => {
    expect(renderer.create(<SocialButton provider="Google" />).toJSON()).toMatchSnapshot();
  });

  it('provides Facebook sign in', () => {
    expect(renderer.create(<SocialButton provider="Facebook" />).toJSON()).toMatchSnapshot();
  });
});
