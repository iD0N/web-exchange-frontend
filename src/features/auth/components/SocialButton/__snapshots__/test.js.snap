// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Social buttons provides Facebook sign in 1`] = `
<a
  className="ant-btn ant-btn-facebook"
  href="https://auth.alpha.crypto.io/oauth2/authorize?redirect_uri=http%3A%2F%2Flocalhost%3A3000%2Foauth2response&response_type=code&client_id=3ik2sk5cufec8mlvbnk3ht9e9d&scope=openid+profile+email+phone+aws.cognito.signin.user.admin&identity_provider=Facebook"
  onClick={[Function]}
>
  <i
    className="anticon anticon-facebook"
  >
    <svg
      aria-hidden="true"
      className=""
      data-icon="facebook"
      fill="currentColor"
      height="1em"
      viewBox="64 64 896 896"
      width="1em"
    >
      <path
        d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-92.4 233.5h-63.9c-50.1 0-59.8 23.8-59.8 58.8v77.1h119.6l-15.6 120.7h-104V912H539.2V602.2H434.9V481.4h104.3v-89c0-103.3 63.1-159.6 155.3-159.6 44.2 0 82.1 3.3 93.2 4.8v107.9z"
      />
    </svg>
  </i>
  <span>
    Login with 
    Facebook
  </span>
</a>
`;

exports[`Social buttons provides Google sign in 1`] = `
<a
  className="ant-btn ant-btn-google"
  href="https://auth.alpha.crypto.io/oauth2/authorize?redirect_uri=http%3A%2F%2Flocalhost%3A3000%2Foauth2response&response_type=code&client_id=3ik2sk5cufec8mlvbnk3ht9e9d&scope=openid+profile+email+phone+aws.cognito.signin.user.admin&identity_provider=Google"
  onClick={[Function]}
>
  <i
    className="anticon anticon-google"
  >
    <svg
      aria-hidden="true"
      className=""
      data-icon="google"
      fill="currentColor"
      height="1em"
      viewBox="64 64 896 896"
      width="1em"
    >
      <path
        d="M881 442.4H519.7v148.5h206.4c-8.9 48-35.9 88.6-76.6 115.8-34.4 23-78.3 36.6-129.9 36.6-99.9 0-184.4-67.5-214.6-158.2-7.6-23-12-47.6-12-72.9s4.4-49.9 12-72.9c30.3-90.6 114.8-158.1 214.7-158.1 56.3 0 106.8 19.4 146.6 57.4l110-110.1c-66.5-62-153.2-100-256.6-100-149.9 0-279.6 86-342.7 211.4-26 51.8-40.8 110.4-40.8 172.4S151 632.8 177 684.6C240.1 810 369.8 896 519.7 896c103.6 0 190.4-34.4 253.8-93 72.5-66.8 114.4-165.2 114.4-282.1 0-27.2-2.4-53.3-6.9-78.5z"
      />
    </svg>
  </i>
  <span>
    Login with 
    Google
  </span>
</a>
`;
